src/functional/geometry/convex_hull.zig

```diff
@@ -0,0 +1,167 @@
+//! https://doi.org/10.1016/0020-0190(72)90045-2.
+//! Convex Hulls are used by the Slide Filter and Optimal PLA algorithms implemented in
+//! `src/functional/swing_slide_filter.zig` and `src/functional/optimal_pla.zig`.
+
+const std = @import("std");
+const mem = std.mem;
+const ArrayList = std.ArrayList;
+const testing = std.testing;
+
+const tersets = @import("../../tersets.zig");
+const Point = tersets.Point;
+
+/// Enum for the angle's `Turn` of three consecutive points A, B, and C. Essentially, it describes
+/// whether the path from A to B to C makes a `left` turn, a `right` turn, or continues in a
+/// straight line also called collinear.
+const Turn = enum(i8) { right, left, collinear };
+
+/// Convex Hull formed by an upper and lower hull.
+pub const ConvexHull = struct {
+    lower_hull: ArrayList(Point),
+    upper_hull: ArrayList(Point),
+
+    // Initialize the container with a given `allocator`.
+    pub fn init(allocator: mem.Allocator) !ConvexHull {
+        return ConvexHull{
+            .lower_hull = ArrayList(Point).init(allocator),
+            .upper_hull = ArrayList(Point).init(allocator),
+        };
+    }
+
+    // Deinitialize the container and free the allocated memory.
+    pub fn deinit(self: *ConvexHull) void {
+        self.lower_hull.deinit();
+        self.upper_hull.deinit();
+    }
+
+    /// Add new `point` to the convex hull.
+    pub fn addPoint(self: *ConvexHull, point: Point) !void {
+        try addPointToHull(&self.upper_hull, Turn.right, point);
+        try addPointToHull(&self.lower_hull, Turn.left, point);
+    }
+};
+
+/// Auxiliary function to add a new `point` to a given `hull` of the convex hull. The function uses
+/// the given `turn` to correctly add the new point.
+fn addPointToHull(hull: *ArrayList(Point), turn: Turn, point: Point) !void {
+    if (hull.items.len < 2) {
+        // The first two points can be add directly.
+        try hull.append(point);
+    } else {
+        var top: usize = hull.items.len - 1;
+        // Remove the last point as long as the `turn` is not the provided.
+        while ((top > 0) and (computeTurn(
+            hull.items[top - 1],
+            hull.items[top],
+            point,
+        ) != turn)) : (top -= 1) {
+            _ = hull.pop();
+        }
+        try hull.append(point);
+    }
+}
+
+/// Compute turn created by the path from `first_point` to `middle_point` to `last_point`.
+fn computeTurn(first_point: Point, middle_point: Point, last_point: Point) Turn {
+    const distance_last_middle: f64 = @floatFromInt(last_point.time - middle_point.time);
+    const distance_middle_first: f64 = @floatFromInt(middle_point.time - first_point.time);
+
+    const cross_product = (middle_point.value - first_point.value) * distance_last_middle -
+        (last_point.value - middle_point.value) * distance_middle_first;
+
+    return if (cross_product == 0) Turn.collinear else if (cross_product > 0) Turn.right else Turn.left;
```

Comment:
Maybe format if each boolean expression and return statement on separate line as the line is currently more than 100 characters.

Reply:
Are we using the same way to measure the size of a line? Are we supposed to count from column zero or the start of the line? In this case from the word `return` on. If the latest, VSC says that it has exactly 100 characters. Regardless, I reformatted this line.

Reply:
Generally when programming line length is counted as the [number of monospace characters per line](https://en.wikipedia.org/wiki/Characters_per_line#In_programming), thus whitespace characters such as spaces are also included. So while there are only 100 printable characters in the line, it is indented by four space characters, making the length of the line 104 characters.